presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -79,31 +91,163 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
 
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
             FilterNode newFilterNode = new FilterNode(
                     idAllocator.getNextId(),
-                    node.getSource(),
+                    childDecorrelationResult.node,
                     ExpressionUtils.combineConjuncts(uncorrelatedPredicates));
 
+            Set<Symbol> symbolsToPropagate = Sets.difference(SymbolsExtractor.extractUnique(correlatedPredicates), ImmutableSet.copyOf(correlation));
             return Optional.of(new DecorrelationResult(
                     newFilterNode,
-                    Sets.difference(SymbolsExtractor.extractUnique(correlatedPredicates), ImmutableSet.copyOf(correlation)),
-                    correlatedPredicates));
+                    Sets.union(childDecorrelationResult.symbolsToPropagate, symbolsToPropagate),
+                    ImmutableList.<Expression>builder()
+                            .addAll(childDecorrelationResult.correlatedPredicates)
+                            .addAll(correlatedPredicates)
+                            .build(),
+                    ImmutableMultimap.<Symbol, Symbol>builder()
+                            .putAll(childDecorrelationResult.correlatedSymbolsMapping)
+                            .putAll(extractCorrelatedSymbolsMapping(correlatedPredicates))
+                            .build(),
+                    childDecorrelationResult.atMostSingleRow));
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitLimit(LimitNode node, Void context)
+        {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent() || node.getCount() == 0) {
+                return Optional.empty();
+            }
+
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
+            if (childDecorrelationResult.atMostSingleRow) {
+                return childDecorrelationResultOptional;
+            }
+
+            if (node.getCount() != 1) {
+                return Optional.empty();
+            }
+
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(node.getOutputSymbols())) {
+                return Optional.empty();
+            }
+
+            // Rewrite to aggregation on symbols which we know are constants and are also part of output
+            AggregationNode aggregationNode = new AggregationNode(
+                    idAllocator.getNextId(),
+                    node.getSource(),
+                    ImmutableMap.of(),
+                    ImmutableList.of(ImmutableList.<Symbol>builder()
+                            .addAll(constantSymbols)
+                            .addAll(node.getOutputSymbols())
```

Comment:
In line 163 you already checked that `node.getOutputSymbols()` are covered with `constantSymbols`, is there any point to concatenate them here?

Reply:
I would say for readability (no what the heck moment), but I'm okey either way